### ISSUE TITLE
Please document: What happens if there is a has collision?

### ISSUE THREAD

**Issue by likebike** [restic/restic#1732]
Please document clearly what happens if there is a hash collision.

I read the existing docs, but I didn't see any mention of this scenario.  I looked at the source code, and I think this code answers the question:

```
// internal/archiver/archiver.go:166

func (arch *Archiver) saveChunk(ctx context.Context, chunk chunker.Chunk, p *restic.Progress, token struct{}, file fs.File, resultChannel chan<- saveResult) {
    defer freeBuf(chunk.Data)

    id := restic.Hash(chunk.Data)
    err := arch.Save(ctx, restic.DataBlob, chunk.Data, id)
    // TODO handle error
    if err != nil {
        debug.Log("Save(%v) failed: %v", id, err)
        fmt.Printf("\nerror while saving data to the repo: %+v\n", err)
        panic(err)
    }
    
    p.Report(restic.Stat{Bytes: uint64(chunk.Length)})
    arch.blobToken <- token
    resultChannel <- saveResult{id: id, bytes: uint64(chunk.Length)}
}
```

...and...

```
// internal/archiver/archiver.go:96

// Save stores a blob read from rd in the repository.
func (arch *Archiver) Save(ctx context.Context, t restic.BlobType, data []byte, id restic.ID) error {
    debug.Log("Save(%v, %v)\n", t, id)
    
    if arch.isKnownBlob(id, restic.DataBlob) {
        debug.Log("blob %v is known\n", id)
        return nil
    }

    _, err := arch.repo.SaveBlob(ctx, t, data, id)
    if err != nil {
        debug.Log("Save(%v, %v): error %v\n", t, id, err)
        return err
    }

    debug.Log("Save(%v, %v): new blob\n", t, id)
    return nil
}
```

So, it seems that if there were a hash collision, the newer data would just be assumed to be the same as the older data, and would therefore not be stored in the repository (silent data loss).  During a restore, the older data would be used instead of the new data, resulting in a corrupted restored file... but at least the corruption would be limited to that one item, and the rest of the restic system will continue to function.

...Is this correct?

Maybe it would be a good idea to add some justification for why we think it is OK to lose data in this scenario -- I guess it's because it happens so infrequently...

Thanks!

**Comment by likebike** [comment]
I found this page to be a good explanation for why we don't really need to worry about a collision:  https://www.reddit.com/r/crypto/comments/6rxgs1/what_is_the_significance_of_a_sha2256_hash/

In particular, I liked this comment by 'atoponce':
```
> Physics also places a lower bound on the required energy of computation.
> I don't have a number. But it's probably ridiculously high, too.

Energy is calculated as

    Energy = kT 

Where 'k' is Boltzmann's constant of 1.38064852×10−16 ergs per Kelvin, and "T" is the
temperature of the system. The background temperature of the Universe is about
2.72548 Kelvin. Running a computer at a lower temperature would require more energy,
thus it would be less efficient.

As a result, this means that to flip a single bit with our ideal computer, it requires:

    Energy = (1.38064852×10^−16 ergs per Kelvin) × (2.72548 Kelvin) = 3.762929928×10^-16ergs 

The total energy output from our star is about 1.2×1034 Joules per year. Because one
Joule is the same as 1×107 ergs, then the total annual energy output of the Sun is about
1.2×1041 ergs. So, doing some basic math:

    Bits flipped = (1.2×10^41 ergs) / (3.762929928×10^-16 ergs per bit) = 3.189004374×10^56 bits 

3.189004374×1056 bits means I can flip every combination of bits in a 187-bit number, if I
could harness 100% of the solar energy output from the sun each year.

Our star is a weak one. Rather, if we harnessed 100% of the energy released in a hypernova,
we could do a lot more work. A hypernova is calculated to release something around
1046 Joules or 1053 ergs of energy. Doing that math:

    Bits flipped = (10^53 ergs) / (3.762929928×10^-16 ergs per bit) = 2.657503608×10^68 bits

2.657503608×1068 bits is approximately 227-bits. This is a computation orgy turned up to 11.

Of course, I would have to harness 100% of that energy into my ideal computer, to flip every
combination of these bits. Never mind finding transportation to get me to that hypernova, the
time taken in that travel (how many millions of light years away is it?), and the cost of the
equipment to harness the released energy.

This strongly implies that is it just not physically possible to flip every combination of bits in a
256-bit number.
```

**Comment by likebike** [comment]
After reading the source code, these are my **guesses** for what happens for various kinds of collisions:

```
    If Hash(Chunk) collides:  One chunk of data would be lost, resulting in corruption
                              of dependent file(s).

    If a Pack ID collides:  The extent of damage depends on the backend.
        * The built-in backends (local, sftp, rest-server) save packs like this:

                  fs.OpenFile(filename, os.O_CREATE|os.O_EXCL|os.O_WRONLY, backend.Modes.File)

          The O_CREATE combined with O_EXCL will prevent overwriting of
          old files.  Therefore, the save operation will produce an error, and
          Restic will probably crash.  The user would then need to try to save
          the data again, and new IDs would be used, and everything would be OK.

        * The built-in cloud backends (azure, ...) also try to use a "Create()" API
          call, to indicate that the backend should not overwrite existing data.

        * 'rclone' backends (for example 'ftp', 'webdav', 'dropbox') use "Update()"
          instead of a "Create" instruction.  Therefore, I think a pack collision
          would cause the old pack to be replaced by the new pack, which would
          cause major, unrecoverable Restic Repository corruption.

    If an Index ID collides:  The 'restic' program would probably crash, but the
                              data in the repository would be fine.  The user could
                              work around the problem by disabling the index, or
                              maybe by re-building the index.

    If a Snapshot ID collides:  Similar to Packs -- the result depends on the backend:
        * If the backend is well-behaved, 'restic' will crash but won't corrupt data,
          and the user will be able to try again and everything will be OK.

        * If the backend uses Update() instead of Create, the previous snapshot
          will be overwritten, which will cause some minor repository corruption.
          The user can work around the issue by deleting corrupted snapshot files.
```

**Comment by fd0** [comment]
Thanks for the in-depth analysis, I think it is correct :)

The only thing that's really bad (because the user wouldn't be notified) is if `Hash(chunk)` collides. The probability for that is extremely low, after all we're using a cryptographic hash function (SHA-256), one requirements for such a function is that it must be really really hard to **produce** a collision **on purpose**. Therefore, it's very unlikely that this happens by accident.

**Comment by rawtaz** [comment]
@likebike You seem to understand these things and have done a great job looking into it. Would you mind writing up a PR for the documentation? I imagine we'd put this information in some of the technical sections, as it's on a technical level that most users won't care about. And it's not a FAQ so it probably doesn't go in there.

**Comment by cd1188** [comment]
Maybe an info in a log would be good if a collision was found or an info by the next check. For example, if the file size is different, but the hash is the same you can save a file "collision" like the "lockfile", then it would be clear. Or is that a performance problem? .. or add a cmd-flag like --paranoid for checking at backup whitch would be a bit slower :) .

I have read that CERN is planing to use restic.. what if there is by some PByte only one collision??

**Comment by aawsome** [comment]
About the comment of @fd0. It really is very, very, very, **very** unlikely to get a hash collision.
Just have a look at the maths and discussions like [this](https://crypto.stackexchange.com/questions/47809/why-havent-any-sha-256-collisions-been-found-yet).

It is in fact so unlikely, that things like the earth getting completely destroyed by an asteroid in the next second is much, much more likely. The probability that any part of your computer (CPU, RAM, whatever) just produces a wrong result out of the blue is much more likely than this.

If we document this, please add a disclaimer that this is only a theoretical discussion and add some references to some of the discussions about hash collision probability.

@cd1188: I do not think that an extra check is needed if we also do not (can not) check for much more likely things where we just assume that things are working.

**Comment by MichaelEischer** [comment]
The probability for a hash collision needs to take the birthday paradox into account. Using the formula `p<=n*(n-1)/2*1/2^m` (m=length of a hash in bits) [1], this gives a probability of 1/2^197 = approx. 5e-60 for 1MB chunks in a 1 PB dataset.  The probability that some disks or your whole computer fails are much much higher than the risk of a hash collision.

Regarding the physical limits on bit flips: There seem to be some [theoretical results](https://physicsworld.com/a/squeezed-thermal-states-could-make-computers-more-energy-efficient/) that could allow for a more energy efficient computation.

[1] https://csyhua.github.io/csyhua/hua-PIEEE.pdf

**Comment by oldium** [comment]
My experience is that when things can go wrong, they will. So if there is a chance that the collision happens, then you simply do not have a reliable backup solution. You do not want to figure out during recovery that the file is corrupted. According to the probability this might happen just for one person on the planet, but this is wrong assumption for a reliable backup solution. Zero probability should be the target. Just my two cents.

**Comment by rawtaz** [comment]
@oldium Please see https://crypto.stackexchange.com/questions/8765/is-there-a-hash-function-which-has-no-collisions . While it would be great, what you want is not practical, so there has to be tradeoffs like with everything else. Also note that Git has the same "problem", yet people use it.

If you are truly concerned about this problem, instead of trying to fix one single software (which again isn't doable), you should instead make sure that you back up your data with two differently working softwares (of which one is restic, of course ;) - then if there were to be a collision in one of them, the other should be fine (although theoretically you could have them collide in the very same place, but it's undescribably unlikely assuming the softwares does hash things differently).

**Comment by aawsome** [comment]
@oldium Sorry to be direct, but you are wrong here. 

> Zero probability should be the target. Just my two cents.

Please tell me how you would achieve this given you have a backup tool with doesn't use hashing? For example, your storage - hard disc, SSD, tape; whatever you take - has a probability greater than zero of flipping random bits. How do you get zero probability of loosing data here?

The point is, all things which are implemented to "secure" data are just measurements to decrease the probability of "things going wrong" down to a ridiculous low value - there is no way for zero probability in real life, sorry to tell you!

**Comment by oldium** [comment]
> Please tell me how you would achieve this given you have a backup tool which doesn't use hashing? For example, your storage - hard disc, SSD, tape; whatever you take - has a probability greater than zero of flipping random bits. How do you get zero probability of loosing data here?

That is a completely valid point here. To overcome storage failure, you can duplicate the storage data - copy it to two different locations. You can work-around this. But what if your _backup software_, which is used to make your copies, is source of the error? You cannot workaround it by duplicating the storage, because you duplicate the error actually.

The problem is not the hashing, the problem is that the hashes are used as the _only_ measure to match identical blocks – without checking if the blocks are really equal. A simple check would give you 0% probability to have an error, the hashes should be there just to speed-up searching for the identical block, not to be the only measure.

And one funny fact: A simple copy to a date-based folder is more reliable than a backup using hashes, when other measures to ensure block equality are not used.

**Comment by aawsome** [comment]
> To overcome storage failure, you can duplicate the storage data - copy it to two different locations.

No, you don't overcome it. Each of your two storages can has a probability greater than zero, that there is a bitflip. The probability that the bitflip occurs on both storages at the exact same position exists and is not zero!
(And, as you hopefully realize, you don't get zero probability, if you add as many copies as you want)

**Comment by oldium** [comment]
> No, you don't overcome it. Each of your two storages has a probability greater than zero, that there is a bitflip. The probability that the bitflip occurs on both storages at the exact same position is not zero!

Correct. But this probability should not be increased by the backup software, shouldn't it?

**Comment by rawtaz** [comment]
@oldium Excuse me, but I have to ask; How do you suggest that the verification that blocks etc are identical is done, in practice? Hashes are used in order to have a fixed (or at least smaller) length of signature than the original data itself has. Since the only way to identify a piece of data with 0% probability of collisions is to use all of it, are you suggesting that every time a block of data has to be identified in restic, all of it is read? That would make things immensely slower. So what are you proposing as an alternative to how it works now?

**Comment by oldium** [comment]
@rawtaz You do not need to verify the colliding block immediately during backup, you can upload it as duplicated block, mark it for verification and deduplicate it as soon as it is verified.

**Comment by rawtaz** [comment]
How will you know which blocks are colliding blocks and which ones are not colliding but just deduplicated?

**Comment by dimejo** [comment]
@oldium:
> Zero probability should be the target.


Quoting from [Stackoverflow](https://stackoverflow.com/a/4014170):
> All harddisks ever produced on Earth can't hold enough 1MB files to get a likelihood of a collision of even 0.01% for SHA-256.

That makes it pretty much zero probability IMHO.

**Comment by oldium** [comment]
@rawtaz During upload of „unseen“ block with existing hash, you assume it could be a collision and create a duplicate for verification. The „unseen“ logic needs to use heuristics for decision if the block was already uploaded, for example if the block is from the same file, same date/time, same position, same size, same hash, than it does not need to go through the verification again.

I know that the probability is pretty much zero. I just think that the software should be able to verify that the collision happened and then it should handle it in a way that does not lead to a lost file. When you verify the backup consistency and it tells you that the file hash does not match, it is already too late. Low probability does not mean that it cannot happen, this is why it is non-zero. For example a 1% probability of failure means that the error happens (in average) once in 100 tries, but it could be the first try out of 100, but also the last one, but also first 100 out of 10000 tries. Be it 1 in 10^100 tries or more, who decided that it cannot be the first one?

**Comment by rawtaz** [comment]
I think you are trying to solve a problem that isn't a problem, in a way that I hardly even understand but that sounds completely messy. Using two different software for backing up your data, if you are this concerned about it's integrity, is a much much better solution :)

**Comment by aawsome** [comment]
@oldium While your goal (zero probability of errors in software) is laudable, it can be never reached. What about software bugs? Bugs in the compiler? Bugs in the OS? Hardware bugs on the system that compiles your software? Hardware bugs on the system that runs your software?

In fact it is always about mitigating risks and lowering error probabilities. This is a question about priorities. When dealing wit hardware issues, I feel your priorites are right. If you fear that the probability of any bug in restic is too high for you, the best mitigation strategy would be IMHO to use another backup tool in addition, as @rawtaz also suggested.

**Comment by pvgoran** [comment]
> @rawtaz During upload of „unseen“ block with existing hash, you assume it could be a collision and create a duplicate for verification. The „unseen“ logic needs to use heuristics for decision if the block was already uploaded, for example if the block is from the same file, same date/time, same position, same size, same hash, than it does not need to go through the verification again.

A "heuristic" is, by definition, a potentially incorrect way of solving a problem. Your heuristics for determining "unseen" blocks can fail with non-zero probability, just like hashes: a file can be modified in such a way that a block inside has different contents but the same hash, and then the file's modification date can be updated to be the same as before. Then the "unseen blocks heuristics" won't be able to detect the change.

You can't have 100% reliable backup algorithm, unless it copies all data each time (or directly compares the new data with the already recorded data).

**Comment by fd0** [comment]
Thanks for all the comments on this issue, I think we're clear how to proceed: add a short section to the manual to document it.

**Comment by oldium** [comment]
> A "heuristic" is, by definition, a potentially incorrect way of solving a problem. Your heuristics for determining "unseen" blocks can fail with non-zero probability, just like hashes: a file can be modified in such a way that a block inside has different contents but the same hash, and then the file's modification date can be updated to be the same as before. Then the "unseen blocks heuristics" won't be able to detect the change.

Not really. With some assumptions, like verifying the whole file hash, this could be doable. When you detect file hash issue, you would need to upload the whole file for deduplication. Moreover, with verifying the file hash together with all block hashes you even lower the risk of not detecting file difference. And solving two hashes failure is like solving problem of two consecutive errors. I just want to solve one error, which is storing the wrong file content in the backup, not that the change was not detected (by the way, even simpler heuristics is used nowadays - just using block hashes for the modified file - isn't it?).

**Comment by pvgoran** [comment]
> Not really. With some assumptions, like verifying the whole file hash, this could be doable. When you detect file hash issue, you would need to upload the whole file for deduplication. Moreover, with verifying the file hash together with all block hashes you even lower the risk of not detecting file difference. And solving two hashes failure is like solving problem of two consecutive errors. I just want to solve one error, which is storing the wrong file content in the backup, not that the change was not detected (by the way, even simpler heuristics is used nowadays - just using block hashes for the modified file - isn't it?).

All these measures can only reduce the probability of having wrong data in the backup, not bring it to zero. You can further classify these failures by dividing them into "failures to detect change", "failures to detect difference due to hash collision" and probably something else, and try to eliminate only specific classes of failure; but at the end of the day it's irrelevant: you'll never bring the final probability to zero without copying and comparing the entire data set. (Note that I'm talking here with certain assumptions in mind, namely: perfectly implemented algorithm that runs on perfectly reliable hardware; backup source is a modern general-purpose filesystem, which can't provide guarantees that its files/directories weren't changed, or report specific changed places.)

All you can do is to reduce the risk of having wrong data in the backup. And I think it's more important to prevent things going wrong in varying and unexpected user configurations/environments (weird file systems, eventually consistent backup storage, unreliable network), rather then trying to combat potential failures that **provably** have extremely low probability of occurring.

Also, you need to take into account the performance cost of all these measures. So it's important to compare the cost and the benefits, and employ sane trade-offs. Very few people will want to copy the entire dataset each time to eliminate, say, a 10^-20 chance of backing up wrong data due to hash failure; in practice, it will prevent people from making backups often enough (or even at all), and thus will actually **greatly increase** their chance of losing data, instead of reducing it.

**Comment by oldium** [comment]
The would be happy if the whole file hash could be used as an additional measure to verify the file content as mentioned in #1620.
